### PR TITLE
Adding pre-commit actions

### PR DESCRIPTION
**Pull request by Micnasr**
Should block commits if they do not follow the repo formatting rules.

#76

### CODE REVIEW

File: docs/docs/firmware/dev-setup.md
Comment (BlakeFreer) [inline docs/docs/firmware/dev-setup.md:1]: These are really good docs, just a few minor changes:

1. The heading level should increment by 1 `#` at a time, so replace `####` with `###` in both subheadings headings.
2. You don't need to `bold` the instructions, so remove the `**` on line 188 and 194

**Review by BlakeFreer — CHANGES_REQUESTED** [review]

File: docs/docs/firmware/dev-setup.md [hunk 1]
```diff
@@ -173,3 +173,28 @@ _If you use a different IDE, consider [adding instructions](../tutorials/site-de
         1. Select the `clangd` extension and open :octicons-gear-24: Extension Settings.
         
         1. Paste the full `clangd` executable path in the "Clangd: Path" setting.
+
+
+## Pre-Commit Setup
+
+To ensure that we run formatting and code checks before pushing their code, we use `pre-commit` hooks. This helps catch issues locally before code is committed.
```
Comment (Micnasr) [inline docs/docs/firmware/dev-setup.md:180]: Todo: fix grammar mistake their -> our

Comment (BlakeFreer) [reply docs/docs/firmware/dev-setup.md:180]: You could rephrase this as

"We use `pre-commit` hooks to run formatting and code checks before code is pushed."

**Review by BlakeFreer — CHANGES_REQUESTED** [review]

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -9,12 +9,11 @@ repos:
...
         stages: [commit]
-  
+    
   - repo: local
     hooks:
       - id: clang-format
-        name: Run clang-format
-        entry: clang-format -i -style=file
+        name: Run git-clang-format
+        entry: "\"C:/Program Files/LLVM/bin/git-clang-format\" --style=file --staged"
```
[the remark this answers is not in the document]
Comment (Micnasr) [reply .pre-commit-config.yaml:17]: For some reason clang format worked normally but for git clang format it does not find the path automatically so I am not really sure how to proceed.

```
PS C:\Users\Michael\OneDrive\Documents\FormulaElectric\Repos\racecar> git commit -m "test"
don't commit to branch...................................................Passed
Run git-clang-format.....................................................Failed
- hook id: clang-format
- exit code: 2

C:\Users\Michael\OneDrive\Documents\FormulaElectric\Repos\racecar>py -3 \Program Files\LLVM\bin\git-clang-format --style=file --staged .pre-commit-config.yaml
C:\Users\Michael\AppData\Local\Programs\Python\Python312\python.exe: can't open file 'C:\\Program': [Errno 2] No such file or directory
```

**Comment by Micnasr** [comment]
I have tested it and it is working now on my end.

**Comment by BlakeFreer** [comment]
Looks good, I can run it on my side. Few comments:

1. As written, this will format all lines of code in any staged file. We should only format the lines which were changed by this commit. Add `--diff $(git merge-base HEAD main)` to the command (taken from `.github/workflows/pr-format-check.yml`)
2. Add instructions for installing clang-format to the markdown doc. It can be installed with `pip` so add `clang-format~=19.0` t the pip install instruction.
3. The pre-commit action formats the code but aborts before committing the new code. The developer has to run another `git  commit` to commit the formatted code. We would like this to all happen in the background.
```bash
$ git commit firmware/test.cc -m "Not formatted"
Run git-clang-format.....................................................Failed
- hook id: clang-format
- exit code: 1
- files were modified by this hook

changed files:
    firmware/new.cc

$ git commit firmware/new.cc -m "formatted, will succeed"
Run git-clang-format.....................................................Passed
[pretest 9c882f2] shouldn't go
 1 file changed, 3 insertions(+)
 create mode 100644 firmware/new.cc
```

This should all happen in 1 `git commit` call. Find a way to run `git clang-format` during the commit process to format the code, but not abort if it causes any changes.

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+      # Prevent committing directly to the main branch
+      - id: no-commit-to-branch
+        args: [--branch, main]
+        stages: [commit]
+    
+  - repo: local
+    hooks:
+      - id: clang-format
+        name: Run git-clang-format
+        entry: git clang-format --style=file --staged
```
Comment (BlakeFreer) [inline .pre-commit-config.yaml:17]: As written, this will format all lines of code in any staged file. We should only format the lines which were changed by this commit. Add `--diff $(git merge-base HEAD main)` to the command (taken from `.github/workflows/pr-format-check.yml`)

File: docs/docs/firmware/dev-setup.md [hunk 4]
```diff
@@ -225,3 +225,46 @@ __Important:__ Create an empty file named `.clangd` in the `firmware/` directory
...
+### Minimum Python Version
+
+To install and use `pre-commit`, ensure that you are using **Python 3.6** or newer.
+
+### Installing Pre-Commit
+
+1. Install `pre-commit` via pip:
+
+    ```bash
+    pip install pre-commit
```
Comment (BlakeFreer) [inline docs/docs/firmware/dev-setup.md:261]: Add instructions for installing clang-format to the markdown doc. It can be installed with `pip` so add `clang-format~=19.0` to the pip install instruction.

File: .pre-commit-config.yaml
Comment (BlakeFreer) [inline .pre-commit-config.yaml:1]: The pre-commit action formats the code but aborts before committing the new code. The developer has to run another `git  commit` to commit the formatted code. We would like this to all happen in the background.
```bash
$ git commit firmware/test.cc -m "Not formatted"
Run git-clang-format.....................................................Failed
- hook id: clang-format
- exit code: 1
- files were modified by this hook

changed files:
    firmware/new.cc

$ git commit firmware/new.cc -m "formatted, will succeed"
Run git-clang-format.....................................................Passed
[pretest 9c882f2] shouldn't go
 1 file changed, 3 insertions(+)
 create mode 100644 firmware/new.cc
```

This should all happen in 1 `git commit` call. Find a way to run `git clang-format` during the commit process to format the code, but not abort if it causes any changes.

**Review by BlakeFreer — CHANGES_REQUESTED** [review]

File: docs/docs/firmware/dev-setup.md (hunk 4 above)
Comment (BlakeFreer) [reply docs/docs/firmware/dev-setup.md:261]: Actually, don't do this. Pre-commit lets you specify "additional dependencies" to install with the hook, so add `clang-format~=19.0` to the yaml file. see <https://pre-commit.com/#config-additional_dependencies>

File: .pre-commit-config.yaml
Comment (Micnasr) [reply .pre-commit-config.yaml:1]: I was not able to do this in a clean way as in the only way it would kinda work is to force the command git add . and git commit -m "" right after. I was thinking about it but would it not be better to at least see what changed in the file incase something does not get formatted properly instead of force committing?

Comment (BlakeFreer) [reply .pre-commit-config.yaml:1]: Yes good point, we will force the developer to add manually re-add the files.

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -14,6 +14,8 @@ repos:
...
       - id: clang-format
         name: Run git-clang-format
-        entry: git clang-format --style=file --staged
-        language: system
-        stages: [commit]
\ No newline at end of file
+        entry: git clang-format --style=file --staged --force
+        language: python
+        stages: [commit]
+        additional_dependencies:
```
Comment (BlakeFreer) [inline .pre-commit-config.yaml:20]: good

**Review by BlakeFreer — APPROVED** [review]

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+      # Prevent committing directly to the main branch
+      - id: no-commit-to-branch
+        args: [--branch, main]
+        stages: [commit]
+    
+  - repo: local
+    hooks:
+      - id: clang-format
+        name: Run git-clang-format
+        entry: git clang-format --style=file --staged --force
```
[the remark this answers is not in the document]
Comment (BlakeFreer) [reply .pre-commit-config.yaml]: removed this
